Add types

TaylorBeeston opened the pull request:
`Closes GH-13`.

This PR adds the `types` directory that contains an `index.d.ts` file that allows TypeScript projects to use this package more easily. Additionally, there are some tests to ensure the correctness of this type definition, and some boilerplate associated with `dtslint`. `package.json` has been updated to include `@types/hast` and `dtslint` as dev dependencies. There is also the addition of the test-types script to run dslint, and index.d.ts is exported as the types file for this project.

This is my second time writing a .d.ts file for a project, so please forgive me if there are any issues. One thing I would like to note is that I'm not 100% how this will work with `s` instead of `h`. I'm fairly certain that it will get the same type definition, which should be flexible enough to work with both functions, but I have been having a hard time finding this information out for certain.

TaylorBeeston committed:
Add types

ChristianMurphy commented on types/index.d.ts:
```diff
@@ -0,0 +1,11 @@
+
+import {Element, Properties, Node} from 'hast'
+
+declare function hastscript(
```
Adding some TSDoc annotations can give additional context in typings aware editors.
Descriptions copied from _readme.md_

```suggestion
/**
 * DSL to create virtual hast trees for HTML or SVG
 *
 * @param selector Simple CSS selector
 * @param properties Map of properties
 * @param children (Lists of) child nodes
 */
declare function hastscript(
```

ChristianMurphy reviewed (approved):
LGTM, thanks @TaylorBeeston! :bow:

wooorm commented:
@ChristianMurphy Any thoughts on the `h` vs. `s` thing from the PR comment?

ChristianMurphy commented on types/hastscript-tests.ts:
```diff
@@ -0,0 +1,5 @@
+import h = require('hastscript')
+
+h()
+h('.bar', {class: 'bar'})
+h('.foo', {class: 'bar'}, h('.baz'))
```
Could we add some additional test cases for
* text and array children
* properties being skipped
* return type
* a case were typing should reject

For example:

```suggestion
h() // $ExpectType Element
h('.bar', {class: 'bar'}) // $ExpectType Element
h('.bar', 'child text') // $ExpectType Element
h('.bar', ['child text']) // $ExpectType Element
h('.foo', {class: 'bar'}, h('.baz')) // $ExpectType Element
h('.foo', {class: 'bar'}, [h('.baz')]) // $ExpectType Element
h('.bar',{class: 'bar'} , 'child text') // $ExpectType Element
h('.bar',{class: 'bar'}, ['child text']) // $ExpectType Element
h(false) // $ExpectError
```

ChristianMurphy commented:
> Any thoughts on the h vs. s thing from the PR comment?

Thanks for drawing my attention to that, I missed it the first read through.
That currently will not be supported, this maps to `index.js`, `svg.js` will not pick up on the typings.
TypeScript is a bit finicky when it comes to files that are not the entrypoint.
It wants the typings names the same, and next to the file.
For example:

```
index.js
index.d.ts
svg.js
svg.d.ts
```

We may need to refactor this to flatten the typings out to the top level to support this.

TaylorBeeston committed:
Add TSDoc annotations

Co-authored-by: Christian Murphy <christian.murphy.42@gmail.com>

TaylorBeeston committed:
Add additional type test cases

Co-authored-by: Christian Murphy <christian.murphy.42@gmail.com>

TaylorBeeston commented:
> We may need to refactor this to flatten the typings out to the top level to support this.

I wonder if we can get around this by changing the `tsconfig` file? I will try and experiment with this.

ChristianMurphy commented on types/index.d.ts:
```diff
@@ -0,0 +1,18 @@
+
+import {Element, Properties, Node} from 'hast'
+
+/**
+ * DSL to create virtual hast trees for HTML or SVG
+ *
+ * @param selector Simple CSS selector
+ * @param properties Map of properties
+ * @param children (Lists of) child nodes
+ */
+declare function hastscript(
+  selector?: string,
+  properties?: Properties,
+  children?: string | Node | Array<string | Node>
+): Element
```
The Travis CI/DTSLint error seems to indicate that this currently can't support `children` as the second param, this may need to be overloaded

```suggestion
/**
 * DSL to create virtual hast trees for HTML or SVG
 *
 * @param selector Simple CSS selector
 * @param properties Map of properties
 * @param children (Lists of) child nodes
 */
declare function hastscript(
  selector?: string,
  children?: string | Node | Array<string | Node>
): Element
declare function hastscript(
  selector?: string,
  properties?: Properties,
  children?: string | Node | Array<string | Node>
): Element
```

TaylorBeeston replied:
That's my bad for hitting the merge button without actually testing it 😅

TaylorBeeston committed:
Add function overload

TaylorBeeston committed:
Proposal to handle s

TaylorBeeston commented:
I'm not sure if this is the best way to do it or not as I'm still a little new to working with Github, but I have pushed the commit `Proposal to handle s` that I think will handle our h vs s problem. I added another type declaration, `svg.d.ts`, and added that to the `tsconfig`. It looks like `dtslint` is correctly picking this up, and causing tests to pass. While I'm not 100% sure this fixes it, I did also update `types` in `package.json` to reference the entire `types` directory, and added `svg.d.ts` to `files`, so I believe this should work.

wooorm commented on package.json:
```diff
@@ -41,7 +44,9 @@
     "space-separated-tokens": "^1.0.0"
   },
   "devDependencies": {
+    "@types/hast": "^2.3.1",
```
I believe we’re putting dependent types in `dependencies` directly now, not in dev-deps.

wooorm commented on package.json:
```diff
@@ -41,7 +44,9 @@
   "devDependencies": {
+    "@types/hast": "^2.3.1",
     "browserify": "^16.0.0",
+    "dtslint": "^3.6.12",
```
```suggestion
    "dtslint": "^3.0.0",
```

ChristianMurphy reviewed (changes requested):
I'm not sure the path trick in tsconfig is working, here's how to test.

inside _hastscript_

```
npm install
npm link .
```

outside _hastscript_ create a new folder with

_index.ts_
```ts
import s = require("hastscript/svg")

s("path")
```

and _tsconfig.json_

```json
{
  "compilerOptions": {
    "lib": ["es2015"],
    "strict": true,
    "baseUrl": "."
  }
}
```

in the new folder run
```
npm link hastscript
```

then test by running

```
npx -p typescript -p ts-node ts-node index.ts
```

(or running `ts-node index.ts` directly if you have it installed globally)

Currently I'm getting

```
Could not find a declaration file for module 'hastscript/svg'. '~/test/node_modules/hastscript/svg.js' implicitly has an 'any' type.
  Try `npm install @types/hastscript` if it exists or add a new declaration (.d.ts) file containing `declare module 'hastscript/svg';`

1 import s = require("hastscript/svg")
                     ~~~~~~~~~~~~~~~~
```

TaylorBeeston commented:
@ChristianMurphy gotcha. Thanks for explaining how to test this! I hadn't thought to use `npm link` like that, and have mostly been testing out changes by installing this as a dependency and editing the files in `node_modules`. I've been doing some looking around and it looks like the only solutions are to ask people to add paths to their `tsconfig`'s if they want to use `s` (yuck), flatten the typing to the top level like you've suggested above, wait for TypeScript to add support for multiple `export`s (microsoft/TypeScript#8305), or to use named exports and break backwards compatibility =/

wooorm commented:
Flattening the types then seems the most reasonable?

ChristianMurphy commented:
> Flattening the types then seems the most reasonable?

Agreed, it's what we've done for other packages that have multiple files that need typings, like https://github.com/syntax-tree/unist-util-is

TaylorBeeston committed:
Move @types/hast to dependencies

TaylorBeeston committed:
Update semver of dtslint to ^3.0.0

TaylorBeeston committed:
Flatten type defintions

TaylorBeeston commented:
I have just pushed out three commits, two of them are the suggestions by @wooorm above regarding `package.json`, and the third flattens out all the type definitions following the example of `unist-util-is`. I tested this out using the `link` trick, and everything made it through!

By the way, [unist-util-is has @types/mdast listed as a dev dependency](https://github.com/syntax-tree/unist-util-is/blob/79a313ffefc427fc08b1f6e9433d13095defed24/package.json#L39), just so you know

ChristianMurphy commented:
> By the way, unist-util-is has @types/mdast listed as a dev dependency, just so you know

That is intentional, it is only used in the tests
https://github.com/syntax-tree/unist-util-is/blob/79a313ffefc427fc08b1f6e9433d13095defed24/unist-util-is-test.ts#L2

ChristianMurphy reviewed (approved):
LGTM, thanks @TaylorBeeston! :bow:

wooorm commented:
Sweet, thanks Taylor! Released!
